Automatically get shorthands in StyleProperties::asTextInternal()

Loirooriol opened the pull request:
#### 07aea488888970b456e3b225643c4c99da2b3648
<pre>
Automatically get shorthands in StyleProperties::asTextInternal()
<a href="https://bugs.webkit.org/show_bug.cgi?id=247688">https://bugs.webkit.org/show_bug.cgi?id=247688</a>

Reviewed by Darin Adler.

StyleProperties::asTextInternal() tries to serialize declarations using
shorthands when possible. The list of shorthand candidates for each
longhand was written manually. This was a big and fragile bunch of code
that was missing various shorthands; other shorthands were only mapped
to some of their longhands but not others, causing bugs.

This patch takes advantage of matchingShorthandsForLonghand() to find
the list of shorthand candidates automatically. Note that the returned
vectors now use a different order, as described in CSSOM. I think this
shouldn&apos;t affect the previous callers of matchingShorthandsForLonghand().

To avoid the risk of breaking something, the shorthands that were not
previously used are now explicitly excluded. So this should just be a
refactoring without side-effects, hence no tests.

The change will make it simpler to fix existing bugs in follow-ups.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::shouldAvoidShorthandForLonghand):
Avoid some shorthands to preserve existing behavior.

(WebCore::StyleProperties::asTextInternal const):
The refactoring happens here.

* Source/WebCore/css/process-css-properties.py:
(GenerationContext):
Reorder vectors returned by matchingShorthandsForLonghand().

Canonical link: <a href="https://commits.webkit.org/256511@main">https://commits.webkit.org/256511@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7c2dd3fbb4bfeea74ba4e0267dea70dceea6b30

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96008 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5257 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105560 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99987 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5367 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34016 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88384 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/101382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101669 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82612 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/30993 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39749 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37427 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20586 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4501 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42022 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43911 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39849 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1725,274 +1794,19 @@ StringBuilder StyleProperties::asTextInternal() const
-                break;
             default:
-                break;
+                for (const StylePropertyShorthand& shorthand : matchingShorthandsForLonghand(propertyID)) {
```
I suggest `auto&` instead of `const StylePropertyShorthand&`.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1725,274 +1794,19 @@ StringBuilder StyleProperties::asTextInternal() const
             default:
-                break;
+                for (const StylePropertyShorthand& shorthand : matchingShorthandsForLonghand(propertyID)) {
+                    CSSPropertyID shorthandPropertyID = shorthand.id();
```
I suggest we just write `shorthand.id()` twice instead of a local variable.

Loirooriol committed:
Automatically get shorthands in StyleProperties::asTextInternal()
https://bugs.webkit.org/show_bug.cgi?id=247688

Reviewed by Darin Adler.

StyleProperties::asTextInternal() tries to serialize declarations using
shorthands when possible. The list of shorthand candidates for each
longhand was written manually. This was a big and fragile bunch of code
that was missing various shorthands; other shorthands were only mapped
to some of their longhands but not others, causing bugs.

This patch takes advantage of matchingShorthandsForLonghand() to find
the list of shorthand candidates automatically. Note that the returned
vectors now use a different order, as described in CSSOM. I think this
shouldn't affect the previous callers of matchingShorthandsForLonghand().

To avoid the risk of breaking something, the shorthands that were not
previously used are now explicitly excluded. So this should just be a
refactoring without side-effects, hence no tests.

The change will make it simpler to fix existing bugs in follow-ups.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::shouldAvoidShorthandForLonghand):
Avoid some shorthands to preserve existing behavior.

(WebCore::StyleProperties::asTextInternal const):
The refactoring happens here.

* Source/WebCore/css/process-css-properties.py:
(GenerationContext):
Reorder vectors returned by matchingShorthandsForLonghand().

Canonical link: https://commits.webkit.org/256511@main
